File: src/bots/teams.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

/**
 * Teams bot — Outgoing Webhook receiver + Incoming Webhook / Graph API sender.
 *
 * Two integration modes (use either or both):
 *
 *  OUTGOING WEBHOOK (receive @-mentions)
 *    Teams admin: Apps → Manage apps → Outgoing Webhooks → Create.
 *    Set the callback URL to https://<your-host>/bots/teams/webhook.
 *    Copy the security token → TEAMS_WEBHOOK_SECRET.
 *    The bot responds immediately to synchronous commands, or replies
 *    "Working on it…" and posts back via Incoming Webhook when done.
 *
 *  INCOMING WEBHOOK (post to channels proactively)
 *    Teams channel: … → Connectors → Incoming Webhook → Configure.
 *    Copy the URL → TEAMS_INCOMING_WEBHOOK_URL (or per-matter via
 *    TEAMS_MATTER_WEBHOOKS = '{"M-001":"https://...","M-002":"..."}').
 *
 * The router (registerTeamsBotRoutes) registers:
 *   POST /bots/teams/webhook       — Outgoing Webhook receiver
 *   POST /bots/teams/notify        — Internal: post to a channel (auth-gated)
 *   POST /bots/teams/matter-link   — Associate a matter with a Teams channel
 *
 * Security:
 *   - HMAC-SHA256 verification on every incoming webhook request.
 *   - The secret token is compared with timing-safe equal.
 *   - Teams channel URLs are validated against the expected hostname.
 */

import { createHmac, timingSafeEqual } from "node:crypto";
import { Config } from "../config.js";
import { logger } from "../logger.js";
import { postToTeamsWebhook } from "../integrations/graph.js";
import { dispatch } from "./dispatcher.js";
import type { Orchestrator } from "../orchestrator.js";
import type { FastifyInstance, FastifyRequest } from "fastify";

// ─── Types ────────────────────────────────────────────────────────────────────

export interface TeamsMatterLink {
  matterNumber: string;
  /** Incoming Webhook URL for this matter's channel */
  webhookUrl: string;
  /** Teams team ID (optional — for Graph API posting) */
  teamId?: string;
  /** Teams channel ID (optional — for Graph API posting) */
  channelId?: string;
}

// In-memory matter → channel map (persisted to the MatterLinkStore if enabled)
const matterLinks = new Map<string, TeamsMatterLink>();

// Load from env — JSON map: {"M-001":"https://...", "M-002":"https://..."}
try {
  const raw = process.env.TEAMS_MATTER_WEBHOOKS;
  if (raw) {
    const parsed = JSON.parse(raw) as Record<string, string>;
    for (const [mn, url] of Object.entries(parsed)) {
      matterLinks.set(mn, { matterNumber: mn, webhookUrl: url });
    }
  }
} catch { /* ignore — no matter webhooks configured */ }

// ─── HMAC verification ────────────────────────────────────────────────────────

function verifyTeamsSignature(body: string, authHeader: string, secret: string): boolean {
  if (!authHeader?.startsWith("HMAC ")) return false;
  const provided = authHeader.slice(5).trim();
  const expected = createHmac("sha256", secret)
    .update(body, "utf8")
    .digest("base64");
  try {
    return timingSafeEqual(
      Buffer.from(provided, "base64"),
      Buffer.from(expected, "base64"),
    );
  } catch {
    return false;
  }
}

// ─── Outgoing response formatter ─────────────────────────────────────────────

function teamsMarkdownCard(text: string): Record<string, unknown> {
  return { type: "message", text };
}

// ─── Route registration ───────────────────────────────────────────────────────

export function registerTeamsBotRoutes(app: FastifyInstance, orch: Orchestrator): void {
  const secret = process.env.TEAMS_WEBHOOK_SECRET ?? "";
  const defaultWebhookUrl = process.env.TEAMS_INCOMING_WEBHOOK_URL ?? "";

  // ── Outgoing Webhook receiver ─────────────────────────────────────────────
  app.post("/bots/teams/webhook", async (req: FastifyRequest, reply) => {
    if (!secret) {
      return reply.status(503).send({ error: "Teams bot not configured" });
    }

    // Teams sends the raw body + Authorization header for HMAC verification
    const rawBody = JSON.stringify(req.body);
    const auth = (req.headers["authorization"] as string) ?? "";
    if (!verifyTeamsSignature(rawBody, auth, secret)) {
      logger.warn("Teams webhook: invalid HMAC signature");
      return reply.status(401).send({ error: "Invalid signature" });
    }

    const body = req.body as Record<string, unknown>;
    const msgText = String(
      (body["text"] as string)
      ?? (body["attachments"] as Array<Record<string, unknown>>)?.[0]?.["content"]
      ?? "",
    ).replace(/<at>[^<]+<\/at>/g, "").trim(); // strip @mention XML

    const senderName = String(
      (body["from"] as Record<string, unknown>)?.["name"] ?? "Unknown",
    );
    const channelData = body["channelData"] as Record<string, Record<string, unknown>> | undefined;
    const channelId = String(channelData?.["channel"]?.["id"] ?? "");
    const teamId = String(channelData?.["team"]?.["id"] ?? "");

    logger.info("Teams bot message", { senderName, command: msgText.slice(0, 60) });

    const response = await dispatch(
      { text: msgText, senderName, channelId, teamId, platform: "teams" },
      orch,
    );

    // If async work is queued, post the deferred result back via webhook
    if (response.asyncWork) {
      const targetUrl = matterLinks.get(channelId)?.webhookUrl ?? defaultWebhookUrl;
      setImmediate(async () => {
        try {
          const result = await response.asyncWork!();
          if (targetUrl) {
            await postToTeamsWebhook(targetUrl, "Big Michael", result);
          }
        } catch (err) {
          logger.error("Teams async work failed", { error: (err as Error).message });
          if (targetUrl) {
            await postToTeamsWebhook(targetUrl, "Big Michael", `Error: ${(err as Error).message}`).catch(() => {});
          }
        }
      });
    }

    // Synchronous response — Teams expects this within 5 s
    return teamsMarkdownCard(response.immediate);
  });

  // ── Internal: post a message to a channel ────────────────────────────────
  app.post("/bots/teams/notify", async (req: FastifyRequest, reply) => {
    const { matterNumber, title, text, webhookUrl, facts } = req.body as {
      matterNumber?: string; title: string; text: string;
      webhookUrl?: string; facts?: Array<{ name: string; value: string }>;
    };

    const url = webhookUrl
      ?? (matterNumber ? matterLinks.get(matterNumber)?.webhookUrl : undefined)
      ?? defaultWebhookUrl;

    if (!url) return reply.status(400).send({ error: "No webhook URL configured for this matter" });

    await postToTeamsWebhook(url, title, text, facts);
    return { ok: true };
  });

  // ── Matter → channel link management ─────────────────────────────────────
  app.post("/bots/teams/matter-link", async (req: FastifyRequest, reply) => {
    const { matterNumber, webhookUrl, teamId, channelId } = req.body as TeamsMatterLink;
    if (!matterNumber || !webhookUrl) {
      return reply.status(400).send({ error: "matterNumber and webhookUrl required" });
    }
    matterLinks.set(matterNumber, { matterNumber, webhookUrl, teamId, channelId });
    logger.info("Teams matter link set", { matterNumber });
    return { ok: true, matterNumber };
  });

  app.get("/bots/teams/matter-link/:matterNumber", async (req: FastifyRequest, reply) => {
    const { matterNumber } = req.params as { matterNumber: string };
    const link = matterLinks.get(matterNumber);
    if (!link) return reply.status(404).send({ error: "No Teams link for this matter" });
    return link;
  });

  app.delete("/bots/teams/matter-link/:matterNumber", async (req: FastifyRequest, reply) => {
    const { matterNumber } = req.params as { matterNumber: string };
    matterLinks.delete(matterNumber);
    return { ok: true };
  });
}

// ─── Proactive: post when a task completes ────────────────────────────────────

/**
 * Attach to orchestrator.progressEmitter to post Teams notifications when
 * a task completes and there is a channel linked to the matter.
 */
export function attachTeamsTaskNotifier(
  orch: Orchestrator,
  defaultWebhookUrl?: string,
): void {
  orch.progressEmitter.on("task:*", async (event: { type: string; data: Record<string, unknown> }, taskId: string) => {
    if (event.type !== "complete") return;
    try {
      const task = await orch.getTask(taskId);
      if (!task?.matterNumber) return;
      const link = matterLinks.get(task.matterNumber);
      const url = link?.webhookUrl ?? defaultWebhookUrl;
      if (!url) return;

      await postToTeamsWebhook(url,
        `Matter ${task.matterNumber} — Task Complete`,
        task.output?.slice(0, 800) ?? "Task completed.",
        [
          { name: "Task ID", value: task.id },
          { name: "Workflow", value: task.workflowType },
          { name: "Findings", value: String(task.findings?.length ?? 0) },
        ],
      );
    } catch (err) {
      logger.warn("Teams task notifier failed", { error: (err as Error).message });
    }
  });
}

// ─── Public API ───────────────────────────────────────────────────────────────

export { matterLinks };

export async function postToMatter(matterNumber: string, title: string, text: string): Promise<void> {
  const link = matterLinks.get(matterNumber);
  const url = link?.webhookUrl ?? process.env.TEAMS_INCOMING_WEBHOOK_URL ?? "";
  if (!url) throw new Error(`No Teams channel linked to matter ${matterNumber}`);
  await postToTeamsWebhook(url, title, text);
}

File: biglaw-go/internal/bots/teams.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

// Teams bot — Outgoing Webhook receiver + Incoming Webhook sender.
// HMAC-SHA256 verification on every incoming request.
// Matter → channel links stored in memory and pre-loaded from
// TEAMS_MATTER_WEBHOOKS = '{"M-001":"https://...","M-002":"..."}' env var.

package bots

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"regexp"
	"strings"
	"sync"

	"github.com/discover-legal/biglaw-go/internal/integrations"
	"github.com/discover-legal/biglaw-go/internal/strutil"
)

// ─── Matter link store ────────────────────────────────────────────────────────

// TeamsMatterLink maps a matter number to a Teams Incoming Webhook URL.
type TeamsMatterLink struct {
	MatterNumber string `json:"matterNumber"`
	WebhookURL   string `json:"webhookUrl"`
	TeamID       string `json:"teamId,omitempty"`
	ChannelID    string `json:"channelId,omitempty"`
}

var (
	teamsMu    sync.RWMutex
	teamsLinks = map[string]TeamsMatterLink{}
)

func init() {
	raw := os.Getenv("TEAMS_MATTER_WEBHOOKS")
	if raw == "" {
		return
	}
	var parsed map[string]string
	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
		return
	}
	for mn, u := range parsed {
		teamsLinks[mn] = TeamsMatterLink{MatterNumber: mn, WebhookURL: u}
	}
}

// SetTeamsMatterLink stores or updates a matter → channel link.
func SetTeamsMatterLink(link TeamsMatterLink) {
	teamsMu.Lock()
	defer teamsMu.Unlock()
	teamsLinks[link.MatterNumber] = link
}

// GetTeamsMatterLink returns the link for a matter (ok=false if not found).
func GetTeamsMatterLink(matterNumber string) (TeamsMatterLink, bool) {
	teamsMu.RLock()
	defer teamsMu.RUnlock()
	l, ok := teamsLinks[matterNumber]
	return l, ok
}

// DeleteTeamsMatterLink removes a matter link.
func DeleteTeamsMatterLink(matterNumber string) {
	teamsMu.Lock()
	defer teamsMu.Unlock()
	delete(teamsLinks, matterNumber)
}

// ─── HMAC verification ────────────────────────────────────────────────────────

var teamsAtRe = regexp.MustCompile(`<at>[^<]+</at>`)

// VerifyTeamsSignature checks the HMAC-SHA256 Authorization header.
func VerifyTeamsSignature(body, authHeader, secret string) bool {
	if !strings.HasPrefix(authHeader, "HMAC ") {
		return false
	}
	provided := strings.TrimSpace(authHeader[5:])
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))
	expected := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(provided), []byte(expected))
}

// ─── Route handlers (framework-agnostic) ─────────────────────────────────────

// TeamsWebhookRequest is the deserialized body of a Teams Outgoing Webhook.
type TeamsWebhookRequest struct {
	Text        string `json:"text"`
	Attachments []struct {
		Content string `json:"content"`
	} `json:"attachments"`
	From struct {
		ID   string `json:"id"`
		Name string `json:"name"`
	} `json:"from"`
	ChannelData struct {
		Channel struct {
			ID string `json:"id"`
		} `json:"channel"`
		Team struct {
			ID string `json:"id"`
		} `json:"team"`
	} `json:"channelData"`
}

// HandleTeamsWebhook processes an incoming Teams Outgoing Webhook request.
// rawBody is the verbatim request body string (for HMAC verification).
// Returns the JSON bytes to send back immediately, or an error.
func HandleTeamsWebhook(rawBody, authHeader string, orch OrchestratorFacade) ([]byte, int, error) {
	secret := os.Getenv("TEAMS_WEBHOOK_SECRET")
	if secret == "" {
		return jsonErr("Teams bot not configured"), 503, nil
	}
	if !VerifyTeamsSignature(rawBody, authHeader, secret) {
		slog.Warn("Teams webhook: invalid HMAC signature")
		return jsonErr("Invalid signature"), 401, nil
	}

	var req TeamsWebhookRequest
	if err := json.Unmarshal([]byte(rawBody), &req); err != nil {
		return jsonErr("bad request"), 400, nil
	}
	text := teamsAtRe.ReplaceAllString(req.Text, "")
	if req.Text == "" && len(req.Attachments) > 0 {
		text = req.Attachments[0].Content
	}
	text = strings.TrimSpace(text)
	channelID := req.ChannelData.Channel.ID
	teamID := req.ChannelData.Team.ID
	if !publicBotCommand(text) && (!idAllowed(os.Getenv("TEAMS_ALLOWED_USER_IDS"), req.From.ID) ||
		!idAllowed(os.Getenv("TEAMS_ALLOWED_TEAM_IDS"), teamID)) {
		slog.Warn("Teams webhook: sender not authorized", "teamId", teamID)
		return jsonErr("Sender is not authorized for Big Michael"), 403, nil
	}

	response := Dispatch(BotMessage{
		Text:       text,
		SenderName: req.From.Name,
		ChannelID:  channelID,
		TeamID:     teamID,
		Platform:   PlatformTeams,
	}, orch)

	if response.AsyncWork != nil {
		targetURL := ""
		if l, ok := GetTeamsMatterLink(channelID); ok {
			targetURL = l.WebhookURL
		}
		if targetURL == "" {
			targetURL = os.Getenv("TEAMS_INCOMING_WEBHOOK_URL")
		}
		if !enqueueBotWork(func() {
			result, err := response.AsyncWork()
			if err != nil {
				result = fmt.Sprintf("Error: %s", err.Error())
			}
			if targetURL != "" {
				if postErr := integrations.PostToTeamsWebhook(targetURL, "Big Michael", result, nil); postErr != nil {
					slog.Warn("Teams async post failed", "error", postErr)
				}
			}
		}) {
			return jsonErr("Big Michael is busy; please retry shortly"), 503, nil
		}
	}

	out, _ := json.Marshal(map[string]string{"type": "message", "text": response.Immediate})
	return out, 200, nil
}

// PostToMatter posts a message to the Teams channel linked to a matter.
func PostToMatter(matterNumber, title, text string) error {
	l, ok := GetTeamsMatterLink(matterNumber)
	if !ok {
		u := os.Getenv("TEAMS_INCOMING_WEBHOOK_URL")
		if u == "" {
			return fmt.Errorf("no Teams channel linked to matter %s", matterNumber)
		}
		return integrations.PostToTeamsWebhook(u, title, text, nil)
	}
	return integrations.PostToTeamsWebhook(l.WebhookURL, title, text, nil)
}

// NotifyTaskComplete posts a completion notification for a task to its linked Teams channel.
func NotifyTeamsTaskComplete(taskID, matterNumber, workflowType, output string, findingCount int) {
	l, ok := GetTeamsMatterLink(matterNumber)
	url := ""
	if ok {
		url = l.WebhookURL
	} else {
		url = os.Getenv("TEAMS_INCOMING_WEBHOOK_URL")
	}
	if url == "" {
		return
	}
	body := output
	if len(body) > 800 {
		body = strutil.Truncate(body, 800)
	}
	facts := []integrations.WebhookFact{
		{Name: "Task ID", Value: taskID},
		{Name: "Workflow", Value: workflowType},
		{Name: "Findings", Value: fmt.Sprintf("%d", findingCount)},
	}
	if err := integrations.PostToTeamsWebhook(url, fmt.Sprintf("Matter %s — Task Complete", matterNumber), body, facts); err != nil {
		slog.Warn("Teams task notifier failed", "error", err)
	}
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

func jsonErr(msg string) []byte {
	b, _ := json.Marshal(map[string]string{"error": msg})
	return b
}
